kostal/integrationpluginkostal.cpp

```diff
@@ -0,0 +1,447 @@
+            qCDebug(dcKostal()) << thing << "battery work capacity changed" << batteryWorkCapacity << "Wh";
+            thing->setStateValue(kostalBatteryCapacityStateTypeId, batteryWorkCapacity / 1000.0); // kWh
+        });
+
+        info->finish(Thing::ThingErrorNoError);
+    }
+}
+
+void IntegrationPluginKostal::postSetupThing(Thing *thing)
+{
+    if (thing->thingClassId() == kostalInverterThingClassId) {
+
+        KostalModbusTcpConnection *kostalConnection = m_kostalConnections.value(thing);
+
+        // Check if we have to create the meter for the Kostal inverter
+        if (myThings().filterByParentId(thing->id()).filterByThingClassId(kostalMeterThingClassId).isEmpty()) {
+
+            qCDebug(dcKostal()) << "--> Read block \"powerMeterValues\" registers from:" << 220 << "size:" << 38;
+            QModbusReply *reply = kostalConnection->readBlockPowerMeterValues();
+            if (reply) {
+                if (!reply->isFinished()) {
+                    connect(reply, &QModbusReply::finished, this, [=](){
+                        if (reply->error() == QModbusDevice::NoError) {
+                            const QModbusDataUnit unit = reply->result();
+                            const QVector<quint16> blockValues = unit.values();
+
+                            bool notZero = false;
+                            for (int i = 0; i < blockValues.size(); i++) {
+                                if (blockValues.at(i) != 0) {
+                                    notZero = true;
+                                    break;
+                                }
+                            }
+
+                            if (notZero) {
+                                qCDebug(dcKostal()) << "There is a meter connected but not set up yet. Creating a meter...";
+                                // No meter thing created for this inverter, lets create one with the inverter as parent
+                                ThingClass meterThingClass = thingClass(kostalMeterThingClassId);
+                                ThingDescriptor descriptor(kostalMeterThingClassId, meterThingClass.name(), QString(), thing->id());
+                                // No params required, all we need is the connection
+                                emit autoThingsAppeared(ThingDescriptors() << descriptor);
+                            } else {
+                                qCDebug(dcKostal()) << "There is no meter connected to the inverter" << thing;
+                            }
+                        }
+                    });
+
+                    connect(reply, &QModbusReply::errorOccurred, this, [reply] (QModbusDevice::Error error){
+                        qCWarning(dcKostal()) << "Modbus reply error occurred while updating block \"powerMeterValues\" registers" << error << reply->errorString();
+                        emit reply->finished();
```

Comment:
Does the ModbusReply not emit finished on error? IIRC it does... The weirdness was just that it may emit errorOccurred 3 times before the finished one, no?

Reply:
fixed